doc/developers/contributing.rst

```diff
@@ -251,11 +251,19 @@ If any of the above seems like magic to you, then look up the `Git documentation
 web, or ask a friend or another contributor for help.
 
 If some conflicts arise between your branch and the ``master`` branch, you need
-to merge ``master``. The command will be::
+to merge ``master``. For that, you first need to fetch the ``upstream``'s
+``master``, and then merge it into your branch::
 
-  $ git merge master
+  $ git fetch upstream
+  $ git merge upstream/master
 
-with ``master`` being synchronized with the ``upstream``.
+In order to synchronize your ``master`` with the ``upstream``'s (assuming you
```

Comment:
I don't understand this. I don't understand the benefit of teaching rebase either.

Reply:
Maybe `git pull upstream master` would be enough?

I'd remove the "(assuming you usually work on branches other than your local's ``master``)" part, since we shouldn't give users the idea that it could be done another way.

Reply:
The rebase is just to leave users with a clean master history on their local copy, which I think is good practice and useful later on when users do `git log` stuff.

The "assuming..." part is kinda to tell users that's how a good practice may look like. Not all users create branches and they may just work on their master. My idea here is to give users an idea of how they can be doing the git-workflow by giving them hints and keywords.

Questions such as https://github.com/scikit-learn/scikit-learn/pull/13496#issuecomment-475940578 are not rare among new contributors, and there's a gap between "here's how you add an origin" and "assuming your master is synced with master" in our documentation.

@jnothman I'm not sure which part is not clear here to you. I'm happy to expand and clarify this part, I was trying to keep the additions minimal.

Reply:
I see. Then what about something like this

*It is recommended to work on temporary branches rather than directly on ``master``. Since these temporary branches are branched off ``master``, it is helpful to keep your local ``master`` up to date with the upstream's ``master``. Do do this, you can...*

Reply:
I like that, change the wording.